### PR TITLE
Dispose llvm::TargetMachines prior to llvm::Context being disposed

### LINKED ISSUE #118462: Random `STATUS_ACCESS_VIOLATION` on aarch64-pc-windows-msvc
### Problem

> [!NOTE]
> This issue only occurs on Windows ARM! (Tested on macOS Sonoma and Ubuntu ARM)

After using `cargo build` command random dependencies throw `STATUS_ACCESS_VIOLATION` error. I'm 100% sure that dependencies I'm trying to build are valid because at every compile attempt different dependency fails e.g. _toml_edit_, _crossbeam-channel_, _actix-router_.

Sometimes none of imported dependencies fail and out of 5 attempts I was able to build project succesfully twice (I deleted `target` folder after each attempt).

Example output:
```
   ...
   Compiling walkdir v2.4.0
   Compiling actix-router v0.5.1
   Compiling cookie v0.16.2
error: could not compile `actix-router` (lib)

Caused by:
  process didn't exit successfully: `rustc --crate-name actix_router --edition=2018 C:\Users\dervex\.cargo\registry\src\index.crates.io-6f17d22bba15001f\actix-router-0.5.1\src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --diagnostic-width=175 --crate-type lib --emit=dep-info,metadata,link -C embed-bitcode=no -C debuginfo=2 --cfg "feature=\"default\"" --cfg "feature=\"http\"" -C metadata=2b7cf52e726e9f7c -C extra-filename=-2b7cf52e726e9f7c --out-dir C:\Users\dervex\Desktop\argon\target\debug\deps -L dependency=C:\Users\dervex\Desktop\argon\target\debug\deps --extern bytestring=C:\Users\dervex\Desktop\argon\target\debug\deps\libbytestring-299f3658ce02b902.rmeta --extern http=C:\Users\dervex\Desktop\argon\target\debug\deps\libhttp-43d74e08691fef63.rmeta --extern regex=C:\Users\dervex\Desktop\argon\target\debug\deps\libregex-91e5fd69638fe8f5.rmeta --extern serde=C:\Users\dervex\Desktop\argon\target\debug\deps\libserde-752549031e501119.rmeta --extern tracing=C:\Users\dervex\Desktop\argon\target\debug\deps\libtracing-dacc595cf0b6c699.rmeta --cap-lints allow` (exit code: 0xc0000005, STATUS_ACCESS_VIOLATION)
warning: build failed, waiting for other jobs to finish...
```

### Steps

1. Add some dependencies (in my case 23)
2. Use `cargo build` command

### Possible Solution(s)

Improve build scripts on Windows ARM?

### Notes

> [!IMPORTANT]
> I'm running **virtualized** Windows 11 using Parallels 19
> Host Machine: M2 Pro Macbook
> Host OS: macOS Sonoma

### Version

```text
cargo 1.74.0 (ecb9851af 2023-10-18)
release: 1.74.0
commit-hash: ecb9851afd3095e988daaa35a48bc7f3cb748e04
commit-date: 2023-10-18
host: aarch64-pc-windows-msvc
libgit2: 1.7.1 (sys:0.18.0 vendored)
libcurl: 8.4.0-DEV (sys:0.4.68+curl-8.4.0 vendored ssl:Schannel)
os: Windows 10.0.22621 (Windows 11 Professional) [64-bit]
```

### PR DESCRIPTION
**Pull request by wesleywiser**
If the TargetMachine is disposed after the Context is disposed, it can lead to use after frees in some cases.

I've observed this happening occasionally on code compiled for aarch64-pc-windows-msvc using `-Zstack-protector=strong` but other users have reported AVs from host aarch64-pc-windows-msvc compilers as well.

I was not able to extract a self-contained test case yet so there is no accompanying test.

Fixes #118462

### CODE REVIEW

File: compiler/rustc_codegen_llvm/src/lib.rs [hunk 1]
```diff
@@ -460,6 +462,7 @@ impl ModuleLlvm {
 impl Drop for ModuleLlvm {
     fn drop(&mut self) {
         unsafe {
+            drop(ManuallyDrop::take(&mut self.tm));
```
Comment (cuviper) [inline compiler/rustc_codegen_llvm/src/lib.rs:465]: Can't this be done in-place?
```suggestion
            ManuallyDrop::drop(&mut self.tm);
```

**Review by Noratrieb — APPROVED** [review]
It should probably have a lifetime then, but let's merge this first, thanks for the quick fix!
I didn't spot this during review...

**Comment by apiraino** [comment]
Beta backport accepted as per compiler team [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202023-11-30/near/405137855)

@rustbot label +beta-accepted

**Comment by mmastrac** [comment]
For future searchers, this may have resolved a bug we were having building `deno_webgpu`:

```    
process didn't exit successfully: `/Users/username/.rustup/toolchains/1.74.0-aarch64-apple-darwin/bin/rustc --crate-name deno_webgpu --edition=2021 ext/webgpu/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --diagnostic-width=169 --crate-type lib --emit=dep-info,metadata,link -C embed-bitcode=no -C debuginfo=2 -C split-debuginfo=unpacked -C metadata=43e905977801a26c -C extra-filename=-43e905977801a26c --out-dir /Users/username/projects/deno/denoland/deno/target/debug/deps -C incremental=/Users/username/projects/deno/denoland/deno/target/debug/incremental -L dependency=/Users/username/projects/deno/denoland/deno/target/debug/deps --extern deno_core=/Users/username/projects/deno/denoland/deno/target/debug/deps/libdeno_core-9ad5927c8a9a916e.rmeta --extern serde=/Users/username/projects/deno/denoland/deno/target/debug/deps/libserde-df2c86549ae96031.rmeta --extern tokio=/Users/username/projects/deno/denoland/deno/target/debug/deps/libtokio-850e2048e0f3074d.rmeta --extern wgpu_core=/Users/username/projects/deno/denoland/deno/target/debug/deps/libwgpu_core-d3940c680f7fe049.rmeta --extern wgpu_types=/Users/username/projects/deno/denoland/deno/target/debug/deps/libwgpu_types-2dd3168ea63e6930.rmeta -C 'link-args=-fuse-ld=lld -weak_framework Metal -weak_framework MetalPerformanceShaders -weak_framework QuartzCore -weak_framework CoreGraphics' -D 'clippy::all' -D 'clippy::await_holding_refcell_ref' -D 'clippy::missing_safety_doc' -D 'clippy::undocumented_unsafe_blocks' --cfg tokio_unstable -L /Users/username/projects/deno/denoland/deno/target/debug/gn_out/obj -L native=/Users/username/projects/deno/denoland/deno/target/debug/build/objc_exception-65792dd8e535de72/out` (signal: 11, SIGSEGV: invalid memory reference)
```

**Comment by Zalathar** [comment]
Sorry for commenting on a years-old PR, but I was investigating the drop-order thing as part of some attempted cleanups, and wanted to add some additional links for the benefit of anyone else diving into this history:

- https://github.com/rust-lang/rust/pull/87570/commits/6a454edce908ab2ef204052c4ca5bad29fa9a30e (in #87570) is an earlier commit that adjusted the drop order to avoid what was presumably the same use-after-free.

**Pull request merged**